packages/client/window/src/EventEmitter.ts

```diff
@@ -0,0 +1,177 @@
+import { nanoid } from "nanoid";
+import { z } from "zod";
+
+export type EventMap = Record<string, z.ZodTypeAny>;
+
+export interface EventEmitterOptions<
+    IncomingEvents extends EventMap = EventMap,
+    OutgoingEvents extends EventMap = EventMap,
+> {
+    incomingEvents?: IncomingEvents;
+    outgoingEvents?: OutgoingEvents;
+}
+
+export interface SendActionOptions<IncomingEvents extends EventMap, R extends keyof IncomingEvents> {
+    timeoutMs?: number;
+    intervalMs?: number;
+    condition?: (data: z.infer<IncomingEvents[R]>) => boolean;
+}
+export type SendActionArgs<
+    IncomingEvents extends EventMap,
+    OutgoingEvents extends EventMap,
+    K extends keyof OutgoingEvents,
+    R extends keyof IncomingEvents,
+> = {
+    event: K;
+    data: z.infer<OutgoingEvents[K]>;
+    responseEvent: R;
+    options?: SendActionOptions<IncomingEvents, R>;
+};
+
+export interface OnActionOptions<OutgoingEvents extends EventMap, R extends keyof OutgoingEvents> {
+    timeoutMs?: number;
+    condition?: (data: z.infer<OutgoingEvents[R]>) => boolean;
+}
+export type OnActionArgs<
+    IncomingEvents extends EventMap,
+    OutgoingEvents extends EventMap,
+    K extends keyof IncomingEvents,
+    R extends keyof OutgoingEvents,
+> =
+    | {
+          event: K;
+          responseEvent: R;
+          callback: (data: z.infer<IncomingEvents[K]>) => z.infer<OutgoingEvents[R]>;
+          options?: OnActionOptions<OutgoingEvents, R>;
+      }
+    | {
+          event: K;
+          options?: OnActionOptions<OutgoingEvents, R>;
+      };
+
+export class EventEmitter<IncomingEvents extends EventMap, OutgoingEvents extends EventMap> {
+    private listeners: Map<string, (message: MessageEvent) => void> = new Map();
+
+    constructor(
+        protected otherWindow: Window,
+        protected targetOrigin: string,
+        protected incomingEvents: IncomingEvents,
+        protected outgoingEvents: OutgoingEvents
+    ) {
+        this.otherWindow = otherWindow;
+        this.targetOrigin = targetOrigin;
+    }
+
+    send<K extends keyof OutgoingEvents>(event: K, data: z.infer<OutgoingEvents[K]>) {
+        const result = this.outgoingEvents[event].safeParse(data);
+        if (result.success) {
+            this.otherWindow?.postMessage({ event, data }, this.targetOrigin);
+        } else {
+            console.error("Invalid data for event", event, result.error);
```

Comment:
This comment is about the error handling strategy, so it applies to the other cases as well. I see that we do:
```typescript
try {
    ...do stuff
} catch (error) {
    console.log(error)
}
```

My question is, should we just log or should we throw the error to be handled by the one implementing this package?

Reply:
hmm for now leaving as is, but you are right that the error handling can be improved